core/main/src/broker/endpoint_broker.rs

```diff
@@ -0,0 +1,598 @@
+};
+
+#[derive(Clone, Debug)]
+pub struct BrokerSender {
+    pub sender: Sender<BrokerRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct BrokerRequest {
+    pub rpc: RpcRequest,
+    pub rule: Rule,
+}
+
+/// BrokerCallback will be used by the communication broker to send the firebolt response
+/// back to the gateway for client consumption
+#[derive(Clone, Debug)]
+pub struct BrokerCallback {
+    pub sender: Sender<BrokerOutput>,
+}
+
+static ATOMIC_ID: AtomicU64 = AtomicU64::new(0);
+
+impl BrokerCallback {
+    /// Default method used for sending errors via the BrokerCallback
+    pub async fn send_error(&self, request: BrokerRequest, error: RippleError) {
+        let value = serde_json::to_value(JsonRpcError {
+            code: JSON_RPC_STANDARD_ERROR_INVALID_PARAMS,
+            message: format!("Error with {:?}", error),
+            data: None,
+        })
+        .unwrap();
+        let data = JsonRpcApiResponse {
+            jsonrpc: "2.0".to_owned(),
+            id: Some(request.rpc.ctx.call_id),
+            error: Some(value),
+            result: None,
+            method: None,
+            params: None,
+        };
+        let output = BrokerOutput { data };
+        if let Err(e) = self.sender.send(output).await {
+            error!("couldnt send error for {:?}", e);
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BrokerContext {
+    pub app_id: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct BrokerOutput {
+    pub data: JsonRpcApiResponse,
+}
+
+impl BrokerOutput {
+    pub fn is_result(&self) -> bool {
+        self.data.result.is_some()
+    }
+
+    pub fn get_event(&self) -> Option<u64> {
+        if let Some(e) = &self.data.method {
+            let event: Vec<&str> = e.split('.').collect();
+            if let Some(v) = event.first() {
+                if let Ok(r) = v.parse::<u64>() {
+                    return Some(r);
+                }
+            }
+        }
+        None
+    }
+}
+
+impl From<CallContext> for BrokerContext {
+    fn from(value: CallContext) -> Self {
+        Self {
+            app_id: value.app_id,
+        }
+    }
+}
+
+impl BrokerSender {
+    // Method to send the request to the underlying broker for handling.
+    pub async fn send(&self, request: BrokerRequest) -> RippleResponse {
+        if let Err(e) = self.sender.send(request).await {
+            error!("Error sending to broker {:?}", e);
+            Err(RippleError::SendFailure)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct EndpointBrokerState {
+    endpoint_map: Arc<RwLock<HashMap<String, BrokerSender>>>,
+    callback: BrokerCallback,
+    request_map: Arc<RwLock<HashMap<u64, BrokerRequest>>>,
+    extension_request_map: Arc<RwLock<HashMap<u64, ExtnMessage>>>,
+    rule_engine: RuleEngine,
+}
+
+impl EndpointBrokerState {
+    pub fn get(tx: Sender<BrokerOutput>, rule_engine: RuleEngine) -> Self {
```

Comment:
`get` implies return of a new or existing instance, but this is really a constructor so should be named `new` or similar.